### ISSUE TITLE
Coefficient Projection for IGA Spaces

### ISSUE THREAD

**Issue by michi002** [mfem/mfem#4328]
Hello,

I am recently dealing with Isogeometric Analysis (IGA) for the issue  [https://github.com/mfem/mfem/issues/4238](https://github.com/mfem/mfem/issues/4238), but I am by far not an expert on IGA.
For one of my plausibilty checks, I wanted to project a coefficient to a gridfunction defined on an IGA mesh.
When trying, I get the "MFEM abort: method is not overloaded' error for 
`virtual void mfem::FiniteElement::Project(mfem::Coefficient&, mfem::ElementTransformation&, mfem::Vector&) const`.
Should we extend the routine to NURBS spaces or is this kind of interpolation not useful in IGA?
As always I am happy to help coding, but I would need some advice. Below is a minimal example of the problem run at the current master version of mfem.

Greetings,
Michael

```
#include <iostream>
#include "mfem.hpp"

int main(int argc, char *argv[]) {

    const char *mesh_file = "../data/square-nurbs.mesh"; //fix the nurbs-mesh to a simple one
    mfem::Mesh mesh(mesh_file, 1);
    //Refine once or twice to get a more useful mesh
    mesh.UniformRefinement();

    //Define the finite element space on the mesh. Here we use the iga space
    mfem::FiniteElementCollection *fec = mesh.GetNodes()->OwnFEC();
    mfem::FiniteElementSpace fes(&mesh, fec);

    //setup a function coefficient to project
    auto u_fun = [](const mfem::Vector &x) { return std::sin(M_PI*x[0])*std::sin(2.*M_PI*x[1]); };
    mfem::FunctionCoefficient u_coeff(u_fun);

    //project to gridfunction (this is where the error happens)
    mfem::GridFunction u_gf(&fes);
    u_gf.ProjectCoefficient(u_coeff);
    return 0;
}
```

**Comment by IdoAkkerman** [comment]
@michi002 In #4326 I started a discussion about this. I am working on a solution.
I have two:

1)  In `GridFunction::ProjectCoefficient` instead of an element level routine do a global (L2) projection.
This is implemented in the branch `hdiv-nurbs` on line 2349. See routine below, in particular the braced part below the first `else`.

2) I am now working on an element level routine that is more akin to a nodel interpolant, as mentioned in #4326.

I think the solution might be something in between solution 1 and 2. Namely a global implementation of somethin akin a nodal interpolant.


```
void GridFunction::ProjectCoefficient(Coefficient &coeff)
{
   DeltaCoefficient *delta_c = dynamic_cast<DeltaCoefficient *>(&coeff);
   DofTransformation * doftrans = NULL;

   if (delta_c == NULL)
   {
      if (fes->GetNURBSext() == NULL)
      {
         Array<int> vdofs;
         Vector vals;

         for (int i = 0; i < fes->GetNE(); i++)
         {
            doftrans = fes->GetElementVDofs(i, vdofs);
            vals.SetSize(vdofs.Size());
            fes->GetFE(i)->Project(coeff, *fes->GetElementTransformation(i), vals);
            if (doftrans)
            {
               doftrans->TransformPrimal(vals);
            }
            SetSubVector(vdofs, vals);
         }
      }
      else
      {
         // Define and assemble linear form
         LinearForm b(fes);
         b.AddDomainIntegrator(new DomainLFIntegrator(coeff));
         b.Assemble();

         // Define and assemble bilinear form
         BilinearForm a(fes);
         a.AddDomainIntegrator(new MassIntegrator());
         a.Assemble();

         // Set solver and preconditioner
         SparseMatrix A(a.SpMat());
         GSSmoother  prec(A);
         CGSolver cg;
         cg.SetOperator(A);
         cg.SetPreconditioner(prec);
         cg.SetRelTol(1e-12);
         cg.SetMaxIter(1000);
         cg.SetPrintLevel(0);

         // Solve and get solution
         *this = 0.0;
         cg.Mult(b,*this);
      }
   }
   else
   {
      real_t integral;

      ProjectDeltaCoefficient(*delta_c, integral);

      (*this) *= (delta_c->Scale() / integral);
   }
}
```

**Comment by michi002** [comment]
@IdoAkkerman 
I now see, that it is not so easy to define an interpolation like in finite elements. However I found some lecture notes talking about interpolation IGA spaces [Lecture Notes](https://numa.jku.at/media/filer_public/3c/2a/3c2a6222-fd4d-4405-b866-e3757698fe8c/lecture_notes_iga.pdf). In chapter 2 they introduce interpolants and quasi-interpolants. We could work on that.

P.S.: Sry for the double post, I overlooked your previous issue. So we can close this one, if no one has anything to add
